Modules/cjkcodecs/multibytecodec.c

```diff
@@ -895,6 +901,80 @@ _multibytecodec_MultibyteIncrementalEncoder_encode_impl(MultibyteIncrementalEnco
     return encoder_encode_stateful(STATEFUL_ECTX(self), input, final);
 }
 
+/*[clinic input]
+_multibytecodec.MultibyteIncrementalEncoder.getstate
+[clinic start generated code]*/
+
+static PyObject *
+_multibytecodec_MultibyteIncrementalEncoder_getstate_impl(MultibyteIncrementalEncoderObject *self)
+/*[clinic end generated code: output=9794a5ace70d7048 input=4a2a82874ffa40bb]*/
+{
+    const char *pendingbuffer = NULL;
+    Py_ssize_t pendingsize;
+    MultibyteIncrementalEncoderState state;
+
+    memset(&state, 0, sizeof(state));
+    if (self->pending != NULL) {
+        pendingbuffer = PyUnicode_AsUTF8AndSize(self->pending, &pendingsize);
+        if (pendingbuffer == NULL) {
+            return NULL;
+        }
+        if (pendingsize > MAXENCPENDING*4) {
+            PyErr_SetString(PyExc_UnicodeError, "pending buffer too large");
+            return NULL;
+        }
+        memcpy(state.pending, pendingbuffer, pendingsize);
+        state.pendingsize = pendingsize;
+    }
+    state.codecstate = self->state;
+
+    return _PyLong_FromByteArray((const unsigned char *)&state,
```

Comment:
The problem I see here, is that the integer that is returned depends on the endianness of the CPU (since `MultibyteIncrementalEncoderState` contains `pendingsize` which is `Py_ssize_t` and `codecstate` with is `MultibyteCodec_State` and in turn contains stuff that depends on CPU endianness). However encoder states should be independent of CPU endianness. The problem with `pendingsize` could be fixed by replacing it with an `unsigned char` (since its value is `2`), but I don't know what could be done about `codecstate`.

Note that switching the encoder state to something that always include a buffered input wouldn't solve this problem since we still have to handle `codecstate`. Furthermore `codecs.BufferedIncrementalEncoder` is somewhat strange as `getstate` does neither always return an integer nor a `(buffer, integer)` tuple. This is somewhat unclear in the definition of incremental encoder, since most encoders don't need it. It's not exactly clear to me from the description at https://en.wikipedia.org/wiki/Extended_Unix_Code why this encoder needs to buffer input.

Reply:
Yes, that is a problem... it would be nice for getstate from a big endian machine to also be acceptable by a little endian machine.
Is it feasible to detect endianness and adjust appropriately? (e.g. always force getstate value to return little endian representation + make setstate convert big endian on big endian machines)

I'm imagining two functions: encoder_state_from_byte_array and encoder_state_to_byte_array.
- encoder_state_from_byte_array casts byte array to MultibyteIncrementalEncoderState and swaps endianness of codecstate and pendingsize if CPU is big endian.
- encoder_state_to_byte_array swaps endianness of codecstate and pendingsize if CPU is big endian and then converts to byte array.

Reply:
Some "if big endian" checks appear to exist in other places, e.g. https://github.com/python/cpython/blob/55edd0c185ad2d895b5d73e47d67049bc156b654/Modules/_io/textio.c#L725

Reply:
True, but that has nothing to do with codec states, however it is good to know that `PY_BIG_ENDIAN` can be used to check endianness.

Note that we might want to use the endianness that leads to a smaller integer value of the encoder state.

Reply:
After looking into it a bit more I don't think the approach of flipping endianness that I described above will work.

[MultiByteCodec_State](https://github.com/python/cpython/blob/master/Modules/cjkcodecs/multibytecodec.h#L19) contains `unsigned char c[8]` which is independent of endianness. The bytes of `MultiByteCodec_State` only need to be reversed if the member being used in an integer type (e.g. `int i`), but there is no way to check this.

An alternative could be to reduce this:

```
typedef union {
    void *p;
    int i;
    unsigned char c[8];
    ucs2_t u2[4];
    Py_UCS4 u4[2];
} MultibyteCodec_State;
```

to this:

```
typedef struct {
    unsigned char c[8];
} MultibyteCodec_State;
```

As far as I can see, only `int i` and `unsigned char c[8]` are used by current codecs. The one usage of `int i` in `_codecs_cn` is trivial (only sets `int i` to 0 or 1) and can be replaced by `c[0]`. Going down this route, it may even make sense to drop `MultibyteCodec_State` and replace it with the `unsigned char` array to prevent someone accidentally adding a new endian-dependent member to it.

Any thoughts? I wouldn't want to make this change without a second opinion. @doerwalter @methane

Reply:
> typedef struct {
>     unsigned char c[8];
> } MultibyteCodec_State;

+1 on this.

Reply:
This sounds good, but I don't know the intricacies of the multibyte codecs.